src/buffer/out/Row.cpp

```diff
@@ -191,8 +347,331 @@ OutputCellIterator ROW::WriteCells(OutputCellIterator it, const til::CoordType i
+{
+    const auto text = GetText();
+    const auto beg = text.begin();
+    const auto end = text.end();
+    auto it = end;
+
+    for (; it != beg; --it)
+    {
+        // it[-1] is safe as `it` is always greater than `beg` (loop invariant).
+        if (til::at(it, -1) != L' ')
+        {
+            break;
+        }
+    }
+
+    // We're supposed to return the measurement in cells and not characters
+    // and therefore simply calculating `it - beg` would be wrong.
+    //
+    // An example: The row is 10 cells wide and `it` points to the second character.
+    // `it - beg` would return 1, but it's possible it's actually 1 wide glyph and 8 whitespace.
+    return gsl::narrow_cast<til::CoordType>(_columnCount - (end - it));
+}
+
+bool ROW::ContainsText() const noexcept
+{
+    const auto text = GetText();
+    const auto beg = text.begin();
+    const auto end = text.end();
+    auto it = beg;
+
+    for (; it != end; ++it)
+    {
+        if (*it != L' ')
+        {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+std::wstring_view ROW::GlyphAt(til::CoordType column) const noexcept
+{
+    auto col = _clampedColumn(column);
+
+    // Safety: col is [0, _columnCount).
+    const auto beg = _uncheckedCharOffset(col);
+    // Safety: col cannot be incremented past _columnCount, because the last
+    // _charOffset at index _columnCount will never get the CharOffsetsTrailer flag.
+    while (_uncheckedIsTrailer(++col))
+    {
+    }
+    // Safety: col is now [1, _columnCount].
+    const auto end = _uncheckedCharOffset(col);
+
+    return { _chars.begin() + beg, _chars.begin() + end };
+}
+
+DbcsAttribute ROW::DbcsAttrAt(til::CoordType column) const noexcept
+{
+    const auto col = _clampedColumn(column);
+
+    auto attr = DbcsAttribute::Single;
+    // Safety: col is [0, _columnCount).
+    if (_uncheckedIsTrailer(col))
+    {
+        attr = DbcsAttribute::Trailing;
+    }
+    // Safety: col+1 is [1, _columnCount].
+    else if (_uncheckedIsTrailer(::base::strict_cast<size_t>(col) + 1u))
+    {
+        attr = DbcsAttribute::Leading;
+    }
+
+    return { attr };
+}
+
+std::wstring_view ROW::GetText() const noexcept
+{
+    return { _chars.data(), _charSize() };
+}
+
+DelimiterClass ROW::DelimiterClassAt(til::CoordType column, const std::wstring_view& wordDelimiters) const noexcept
+{
+    const auto col = _clampedColumn(column);
+    // Safety: col is [0, _columnCount).
+    const auto glyph = _uncheckedChar(_uncheckedCharOffset(col));
+
+    if (glyph <= L' ')
+    {
+        return DelimiterClass::ControlChar;
+    }
+    else if (wordDelimiters.find(glyph) != std::wstring_view::npos)
+    {
+        return DelimiterClass::DelimiterChar;
+    }
+    else
+    {
+        return DelimiterClass::RegularChar;
+    }
+}
+
+template<typename T>
+constexpr uint16_t ROW::_clampedUint16(T v) noexcept
+{
+    return static_cast<uint16_t>(std::max(T{ 0 }, std::min(T{ 65535 }, v)));

```

Comment:
why 65535? Probably worth adding a comment for that one.

Reply:
fwiw it's the maximum uint16